requirements/pip-compile-superset.py

```diff
@@ -0,0 +1,153 @@
+#!/usr/bin/env python
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import subprocess
+import sys
+from shutil import copyfile
+
+import click
+
+BASE_REQS = "requirements/base.txt"
+DEV_REQS = "requirements/development.txt"
+
+DEV_EXTRAS = [
```

Comment:
Doesn't this CLI essentially add another layer that we need to work though? I would have thought that the old `compile-multi` would have just had a include file for these deps, no? I'm a little concerned that folks will need to come and update this python file whenever they need to tune a dep, so you'd need to add here AND in the `toml`, right?

Reply:
Yes, originally I just had a bash scripts that would run `pip-compile --extra {whole_list_here}`, as we need those semantics somewhere. Those semantics used to live in `.in` files and needed a new place.

Now I realized I needed a few things on tip of the bash script and thought a tiny CLI would be useful so that :
- I wanted a way to make sure `development.txt` is in sync with `base.txt` for the librairies that they share, and prevent any drift. In the past, that was insured by having the `.txt` reference one another. I think I should add a CI check that runs `pip-compile-superset.py compare-versions` to make sure things don't drift for whatever reason. Happy to add this check part of this PR
- `compile-deps` this subcommand should insure that consistency when changing inputs and/or bumping libs. I mean you could run both pip-compile manually. Basically that means we can update both environments pins atomically
- I wanted to offer a way to expand upon what's here. Sure you could `pip install -r requirements/base.txt` and `pip install` whatever else on top, but I wanted to offer a way to pip-compile your stuff alongside what's in here. I added docs as part of this PR both for CONTRIBUTING.md and in the producitonizing-related docs.

Reply:
Makes sense, but I think we should try to keep a single source of truth for this list. Would it not suffice to add it to the `toml` file under the `development` "optional" req's? Just trying to avoid the need to update dependency lists in multiple locations going forward. If there end up being multiple requirement sources in the toml file, perhaps this CLI would perform a linting function which would ensure version consistency.

Reply:
Hey, clarifying here, but the list of `--extra` here is referencing name GROUPs of requirements as specified in the  `pyproject.toml`, these groups are useful in their own rights as someone might want to pick on or compose a collection of those. Those are not individual libs, but named groups of libs defined in `pyproject.toml`.

So this little array represent the list of groups we want to compose into our `requirements/development.txt`. Now why not having one big group called `development` with everything? Well people in the community might want to install different subsets of these groups, say `mysql`, `snowflake` and `trino`. Another company may want `redshift` and `hive`.

To define our own development.txt, I picked everything that's needed in some form or fashion by CI, plus the stuff used in common dev workflows filed under a `development` group. This contains libs like pylint / docker / mypy ...